keras_cv/layers/object_detection/roi_generator.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Mapping
+from typing import Optional
+from typing import Tuple
+
+import tensorflow as tf
+
+from keras_cv import bounding_box
+
+
+class ROIGenerator(tf.keras.layers.Layer):
+    """
+    Generates region of interests (ROI, or proposal) from scores.
+
+    This works for a multi-level input, both boxes and scores are dictionary
+    inputs with the same set of keys.
+
+    Users can configure topk and threshold differently in train and inference.
+
+    Users can choose to combine all levels if nms across all levels are desired.
+
+    The following steps are applied to pair of (boxes, scores):
+    1) pre_nms_topk scores and boxes sorted and selected per level
+    2) nms applied and selected post_nms_topk scores and rois per level
+    3) combined scores and rois across all levels
+    4) post_nms_topk scores and rois sorted and selected
+
+    Args:
+        bounding_box_format: a case-insensitive string which is one of `"xyxy"`,
+            `"rel_xyxy"`, `"xyWH"`, `"center_xyWH"`, `"yxyx"`, `"rel_yxyx"`. The
+            position and shape of the bounding box will be followed by the class and
+            confidence values (in that order). This is required for proper ranking of
+            the bounding boxes. Therefore, each bounding box is defined by 6 values.
+            For detailed information on the supported format, see the
+            [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+        pre_nms_topk_train: int. number of top k scoring proposals to keep before applying NMS in training mode.
+            When RPN is run on multiple feature maps / levels (as in FPN) this number is per
+            feature map / level.
+        nms_score_threshold_train: float. score threshold to use for NMS in training mode.
```

Comment:
so almost all of these parameters are passed into a NMS layer.

My intuition is that wrapping the underlying op (in this case tf.image.combine_non_max_suppression_padded) in a layer and accepting a layer to the constructor instead will produce the best API.

We did this for: https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/object_detection/non_max_suppression.py and it still supports full customization, but it eliminates these length constructor calls - and separates the customization into a logical block.

So in this case, you'd take two layers: a pre-nms-layer and a test-nms layer.

WDYT?

Reply:
Hmmm...not sure if I understand the comment -- the object_detection/non_max_suppression also takes arguments that control nms: `confidence_threshold`, `iou_threshold`, `max_detections`, `max_detections_per_class`, isn't it the same here?

If we do think this is too much, we can follow an established pattern which is:
`nms_iou_threshold`: List[float] -- a list of size 2 floats, the first for train, the second for test.
So we have 4, instead of 8 arguments.

I choose the existing approach mainly that I want to allow a simpler tuner integration, i.e., users can tune train and test configs separately. But I'm ok if we combine them into list.

WDYT?

Reply:
> Hmmm...not sure if I understand the comment -- the object_detection/non_max_suppression also takes arguments that control nms: `confidence_threshold`, `iou_threshold`, `max_detections`, `max_detections_per_class`, isn't it the same here?
> 
> If we do think this is too much, we can follow an established pattern which is: `nms_iou_threshold`: List[float] -- a list of size 2 floats, the first for train, the second for test. So we have 4, instead of 8 arguments.
> 
> I choose the existing approach mainly that I want to allow a simpler tuner integration, i.e., users can tune train and test configs separately. But I'm ok if we combine them into list.
> 
> WDYT?

> Hmmm...not sure if I understand the comment -- the object_detection/non_max_suppression also takes arguments that control nms: confidence_threshold, iou_threshold, max_detections, max_detections_per_class, isn't it the same here?


it is the same, I'm proposing we accept an actual NMS layer as a parameter and produce a default once instead of having all of these prefixed layers.

Reply:
(so you'd take two layers instead of a list of two floats)

Reply:
Here is an example of a comparable API:

https://github.com/keras-team/keras-cv/blob/master/keras_cv/layers/object_detection/nms_prediction_decoder.py#L35

we take a suppression_layer instead of a list of all of the parameters to pass to the layer.

Reply:
OIC. This makes sense. Though for this particular layer, I'd rather not accept *any* nms layer as I want to make sure this works on GPU/TPU -- NMS is very tricky to work across platforms.

Reply:
I see.  Maybe we add a TODO to accept a padded NMS layer then?

My main concern with this thread is the API surface.  Accepting this huge list of nms- prefixed parameters is a bit hard to read - but shouldn't be considered blocking.